File: javascript/nodes/check_m3u_file.js
import { Next } from '../ambler.js';
import fs from 'fs';

const defaultUtils = { stat: (path) => fs.statSync(path) };

export function checkM3UFile(edges, utils = defaultUtils) {
    return async (state) => {
        const { m3uFilePath } = state;

        if (m3uFilePath) {
            try {
                const fileInfo = utils.stat(m3uFilePath);
                if (fileInfo.isFile() && m3uFilePath.endsWith('.m3u')) {
                    return new Next(edges.onRead, state);
                }
            } catch (error) {
                if (error.code === 'ENOENT') {
                    console.log(`File not found: ${m3uFilePath}`);
                } else {
                    console.log(`Error accessing file: ${error.message}`);
                }
            }
        }

        return new Next(edges.onPrompt, state);
    };
}

File: ruby/nodes/check_m3u_file.rb
# frozen_string_literal: true

require_relative '../lib/ambler'

module CheckM3uFile
  DEFAULT_UTILS = { stat: File.method(:stat) }.freeze

  def self.node(edges, utils = DEFAULT_UTILS)
    ->(state) {
      m3u_file_path = state[:m3u_file_path]

      if m3u_file_path
        begin
          stat = utils[:stat].call(m3u_file_path)
          if stat.file? && m3u_file_path.end_with?('.m3u')
            return Ambler::Next.new(edges[:on_read], state)
          end
        rescue Errno::ENOENT
          puts "File not found: #{m3u_file_path}"
        rescue => e
          puts "Error accessing file: #{e.message}"
        end
      end

      Ambler::Next.new(edges[:on_prompt], state)
    }
  end
end
